src/shared/telemetry/awsTelemetryOptOut.ts

```diff
@@ -0,0 +1,95 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+import * as vscode from 'vscode'
+import * as nls from 'vscode-nls'
+import { ext } from '../extensionGlobals'
+import { SettingsConfiguration } from '../settingsConfiguration'
+import { TelemetryService } from './telemetryService'
+const localize = nls.loadMessageBundle()
+
+export class AwsTelemetryOptOut {
+    private static readonly AWS_TELEMETRY_KEY = 'telemetry'
+    private static readonly TELEMETRY_OPT_OUT_SHOWN = 'awsTelemetryOptOutShown'
+
+    private readonly responseYes: string = localize('AWS.generic.response.yes', 'Yes')
+    private readonly responseNo: string = localize('AWS.generic.response.no', 'No')
+
+    public constructor(
+        public readonly service: TelemetryService,
+        private readonly settings: SettingsConfiguration
+    ) {
+        vscode.workspace.onDidChangeConfiguration(async event => {
+            // if telemetry.enableTelemetry changed and user has not expressed a preference
+            if (event.affectsConfiguration('telemetry.enableTelemetry')
+                && this.settings.readSetting(AwsTelemetryOptOut.AWS_TELEMETRY_KEY) === undefined
+            ) {
+                await this.updateTelemetryConfiguration(undefined)
+            }
+        })
+    }
+
+    public async updateTelemetryConfiguration(response: string | undefined) {
+        switch (response) {
+            case this.responseYes:
+                await this.settings.writeSetting(
+                    AwsTelemetryOptOut.AWS_TELEMETRY_KEY,
+                    true,
+                    vscode.ConfigurationTarget.Global
+                )
+                this.service.telemetryEnabled = true
+                break
+
+            case this.responseNo:
+                await this.settings.writeSetting(
+                    AwsTelemetryOptOut.AWS_TELEMETRY_KEY,
+                    false,
+                    vscode.ConfigurationTarget.Global
+                )
+                this.service.telemetryEnabled = false
+                break
+
+            default:
+                await this.settings.writeSetting(
+                    AwsTelemetryOptOut.AWS_TELEMETRY_KEY,
+                    undefined,
+                    vscode.ConfigurationTarget.Global
+                )
+                const vsCodeTelemetryEnabled = this.getVSCodeTelemetrySetting()
+                this.service.telemetryEnabled = vsCodeTelemetryEnabled
+                break
+        }
+    }
+
+    public async ensureUserNotified(): Promise<void> {
+        if (!ext.context.globalState.get<boolean>(AwsTelemetryOptOut.TELEMETRY_OPT_OUT_SHOWN)) {
+            const response = await this.showNotification()
+            await this.updateTelemetryConfiguration(response)
```

Comment:
The code is held up here until a button is pressed or the message is cancelled. A quirk of `showInformationMessage` is that if you press `Esc`, the message gets queued up in the bottom right of VS Code, in the bell icon. The code remains awaiting for a response, which the user may never do. This would hold up the rest of extension activation.

I recommend wrapping line 71 and 72 inside a promise and not awaiting it, as long as the rest of the system will properly work if this wrapped code is deferred.

Reply:
This has been bubbled out to a non-awaited call in extension.ts lines 65-69. Should that be moved here?

Reply:
Brilliant! I overlooked that. I think we're okay here. Please put a function comment on `
ensureUserNotified` along the lines of "Caution: this method awaits a showInfo call, until the user presses a button or cancels the dialog. Using Esc on the dialog will leave the code awaiting. Ensure calling functions handle this suitably, or do not await this method.